File: src/main/java/br/com/brforgers/ducts/inventories/DuctInventory.java
package br.com.brforgers.ducts.inventories;

import br.com.brforgers.ducts.Ducts;
import br.com.brforgers.ducts.blockentities.DuctBlockEntity;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.Slot;
import net.minecraft.world.item.ItemStack;

public class DuctInventory extends AbstractContainerMenu
{
	private DuctBlockEntity duct;
	
	public DuctInventory(int id, Inventory playerInventory, FriendlyByteBuf packet)
	{
		this(id, playerInventory, playerInventory.player.level.getBlockEntity(packet.readBlockPos()) instanceof DuctBlockEntity ductBlock ? ductBlock : null);
	}

	public DuctInventory(int pContainerId, Inventory playerInventory, DuctBlockEntity inventory) 
	{
		super(Ducts.DUCT_MENU.get(), pContainerId);
		
		duct = inventory;
        checkContainerSize(inventory, 1);
        inventory.startOpen(playerInventory.player);
        
        addSlot(new Slot(inventory, 0, 80, 20));
        
        for (int l = 0; l < 3; l++)
        {
        	for (int k = 0; k < 9; k++)
        	{
        		addSlot(new Slot(playerInventory, k + l * 9 + 9, 8 + k * 18, l * 18 + 51));
        	}
        }
        for (int i = 0; i < 9; i++)
        {
        	addSlot(new Slot(playerInventory, i, 8 + i * 18, 109));
        }
	}

	@Override
	public ItemStack quickMoveStack(Player pPlayer, int pIndex) 
	{
		ItemStack itemstack = ItemStack.EMPTY;
		Slot slot = slots.get(pIndex);
		
		if (slot.hasItem())
		{
			ItemStack itemstack1 = slot.getItem();
			itemstack = itemstack1.copy();
			
			if (pIndex < duct.getContainerSize())
			{
				if (!moveItemStackTo(itemstack1, duct.getContainerSize(), slots.size(), true))
					return ItemStack.EMPTY;
			}
			else if (!moveItemStackTo(itemstack1, 0, duct.getContainerSize(), false))
			{
				return ItemStack.EMPTY;
			}
			
			if (itemstack1.isEmpty())
				slot.set(ItemStack.EMPTY);
			else
				slot.setChanged();
		}
		
		return itemstack;
	}

	@Override
	public boolean stillValid(Player pPlayer) 
	{
		return duct.stillValid(pPlayer);
	}

	@Override
	public void removed(Player player) 
	{
		super.removed(player);
        duct.stopOpen(player);
	}
}

File: src/main/kotlin/br/com/brforgers/mods/ducts/inventories/DuctInventory.kt
package br.com.brforgers.mods.ducts.inventories

import br.com.brforgers.mods.ducts.Ducts
import br.com.brforgers.mods.ducts.blockentities.DuctBlockEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.inventory.container.Container
import net.minecraft.inventory.container.Slot
import net.minecraft.item.ItemStack
import net.minecraft.util.math.BlockPos
import javax.annotation.Nonnull


class DuctInventory(syncId: Int, playerInventory: PlayerInventory, inventory: DuctBlockEntity) :
    Container(Ducts.DUCT_MENU, syncId) {
    val duct: DuctBlockEntity

    init {
        duct = inventory
        checkContainerSize(inventory, 1)
        inventory.startOpen(playerInventory.player)
        addSlot(Slot(inventory, 0, 80, 20))
        for (l in 0..2) {
            for (k in 0..8) {
                addSlot(Slot(playerInventory, k + l * 9 + 9, 8 + k * 18, l * 18 + 51))
            }
        }
        for (i1 in 0..8) {
            addSlot(Slot(playerInventory, i1, 8 + i1 * 18, 109))
        }
    }

    constructor(id: Int, playerInventory: PlayerInventory, pos: BlockPos?) : this(
        id,
        playerInventory,
        playerInventory.player.level.getBlockEntity(pos!!) as DuctBlockEntity
    )

    override fun stillValid(@Nonnull player: PlayerEntity): Boolean {
        return duct.stillValid(player)
    }

    @Nonnull
    override fun quickMoveStack(@Nonnull player: PlayerEntity, index: Int): ItemStack {
        var itemstack = ItemStack.EMPTY
        val slot = slots[index]
        if (slot.hasItem()) {
            val itemstack1 = slot.item
            itemstack = itemstack1.copy()
            if (index < duct.containerSize) {
                if (!moveItemStackTo(itemstack1, duct.containerSize, slots.size, true)) {
                    return ItemStack.EMPTY
                }
            } else if (!moveItemStackTo(itemstack1, 0, duct.containerSize, false)) {
                return ItemStack.EMPTY
            }
            if (itemstack1.isEmpty) {
                slot.set(ItemStack.EMPTY)
            } else {
                slot.setChanged()
            }
        }
        return itemstack
    }

    override fun removed(@Nonnull player: PlayerEntity) {
        super.removed(player)
        duct.stopOpen(player)
    }


}